lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp

```diff
@@ -635,76 +627,81 @@ bool TypeLoweringVisitor::lowerProducer(
-      // op. This only supports Bundle or vector of ground type elements.
-      // Recursive aggregate types are not yet supported.
-
-      // This builder ensures that the aggregate construction happens at the
-      // user location, and the LowerTypes algorithm will not touch them any
-      // more, because LowerTypes was reverse iterating on the block and the
-      // user has already been processed.
-      ImplicitLocOpBuilder b(user->getLoc(), user);
-      // Cat all the field elements.
-      Value accumulate;
-      for (auto v : mapping) {
-        if (!v.getType().cast<FIRRTLBaseType>().isGround()) {
-          user->emitError("cannot handle an opaque user of aggregate types "
-                          "with non-ground type elements");
-          return;
-        }
-        if (isa<FVectorType>(val.getType()))
-          accumulate =
-              (accumulate ? b.createOrFold<CatPrimOp>(v, accumulate) : v);
-        else
-          // Bundle subfields are filled from MSB to LSB.
-          accumulate =
-              (accumulate ? b.createOrFold<CatPrimOp>(accumulate, v) : v);
-      }
-      // Cast it back to the original aggregate type.
-      auto input = b.createOrFold<BitCastOp>(val.getType(), accumulate);
-      user->replaceUsesOfWith(val, input);
-    }
+  for (auto *user : llvm::make_early_inc_range(val.getUsers())) {
+    TypeSwitch<Operation *, void>(user)
+        .Case<SubindexOp>([mapping](SubindexOp sio) {
+          Value repl = mapping[sio.getIndex()];
+          sio.replaceAllUsesWith(repl);
+          sio.erase();
+        })
+        .Case<SubfieldOp>([mapping](SubfieldOp sfo) {
+          // Get the input bundle type.
+          Value repl = mapping[sfo.getFieldIndex()];
+          sfo.replaceAllUsesWith(repl);
+          sfo.erase();
+        })
+        .Case<RefSubOp>([mapping](RefSubOp refSub) {
+          Value repl = mapping[refSub.getIndex()];
+          refSub.replaceAllUsesWith(repl);
+          refSub.erase();
+        })
+        .Case<RefForceOp, RefForceInitialOp, RefSendOp, VectorCreateOp,
+              BundleCreateOp>([&](auto op) {
+          // Reconstruct the aggregate for the ref operation.
+          // (or vector/bundle create that were inserted.. woof.)
+          ImplicitLocOpBuilder b(user->getLoc(), user);
+
+          Value input =
+              TypeSwitch<Type, Value>(val.getType())
+                  .template Case<FVectorType>([&](auto vecType) {
+                    return b.createOrFold<VectorCreateOp>(vecType, mapping);
+                  })
+                  .template Case<BundleType>([&](auto bundleType) {
+                    return b.createOrFold<BundleCreateOp>(bundleType, mapping);
+                  })
+                  .Default([&](auto _) -> Value { return {}; });
+          if (!input) {
+            user->emitError("unable to reconstruct source of type ")
+                << val.getType();
+            return;
+          }
```

Comment:
I think we can use for `Default` as well. `Default` currently creates aggrgates by bitcast operations (since there was no vector/bundle_create at that time) so it would be way better to explicitly use vector/bundle_create.

Reply:
As an aside: I've noticed playing with aggregates that often names become `_GEN` -- if we're exploding then reconstructing that would sure explain that.  Unfortunately this doesn't (easily) lend itself to addressing that, but it'd be neat if so.

In your example you're showing another case where we "unevenly" lower things--here, the scalarization of the top module causes the vectors to be split and we'd prefer a vector/bundle create to recover those pieces vs bitcast.
Is that right?
(Makes lots of sense to me!)

Interesting point!! I am unsure... okay digging a little it appears this code was written before we had aggregate creation ops :smile: .
(( Depending on how you look at the purpose of LowerTypes, I was concerned the mechanism for materializing the aggregate might be important -- for claims like "No FIRRTL op will still take an aggregate" or if the story for foreign ops made more sense giving it an opaque blob of bits directly.  In that sense this was a little suboptimal but I didn't have a better story other than rejecting. ))

Anyway a few conversations later I think this makes sense.  I'll change it separately.

cc #3989 which might be useful to revive here re:neater IR but not 100% as haven't gone through that carefully.